drivers/ethernet/eth_nxp_enet_qos/eth_nxp_enet_qos_mac.c

```diff
@@ -0,0 +1,606 @@
+
+	fragment = pkt->frags;
+	tx_desc_ptr->read.control2 |= FIRST_TX_DESCRIPTOR_FLAG;
+	for (int i = 0; i < frags_count; i++) {
+		net_pkt_frag_ref(fragment);
+
+		tx_desc_ptr->read.buf1_addr = (uint32_t)fragment->data;
+		tx_desc_ptr->read.control1 = FIELD_PREP(0x3FFF, fragment->len);
+		tx_desc_ptr->read.control2 |= FIELD_PREP(0x7FFF, total_bytes) | BIT(31);
+
+		fragment = fragment->frags;
+		tx_desc_ptr++;
+	}
+	last_desc_ptr = tx_desc_ptr - 1;
+	last_desc_ptr->read.control1 |= FIELD_PREP(0x80000000, 0b1);
+	last_desc_ptr->read.control2 |= LAST_TX_DESCRIPTOR_FLAG;
+
+	LOG_DBG("Starting TX DMA on packet %p", pkt);
+
+	base->DMA_CH[0].DMA_CHX_TXDESC_RING_LENGTH = frags_count - 1;
+	base->DMA_CH[0].DMA_CHX_TXDESC_TAIL_PTR =
+		ENET_QOS_REG_PREP(DMA_CH_DMA_CHX_TXDESC_TAIL_PTR, TDTP,
+			ENET_QOS_ALIGN_ADDR_SHIFT((uint32_t) tx_desc_ptr));
+
+	return 0;
+}
+
+static void tx_dma_done(struct k_work *work)
+{
+	struct nxp_enet_qos_tx_data *tx_data =
+		CONTAINER_OF(work, struct nxp_enet_qos_tx_data, tx_done_work);
+	struct nxp_enet_qos_mac_data *data =
+		CONTAINER_OF(tx_data, struct nxp_enet_qos_mac_data, tx);
+	struct net_pkt *pkt = tx_data->pkt;
+	struct net_buf *fragment = pkt->frags;
+
+	LOG_DBG("TX DMA completed on packet %p", pkt);
+
+	while (fragment != NULL) {
+		net_pkt_frag_unref(fragment);
+		fragment = fragment->frags;
+	}
+
+	net_pkt_frag_unref(tx_data->tx_header);
+	net_pkt_unref(pkt);
+
+	eth_stats_update_pkts_tx(data->iface);
+
+	k_sem_give(&tx_data->tx_sem);
+}
+
+static enum ethernet_hw_caps eth_nxp_enet_qos_get_capabilities(const struct device *dev)
+{
+	return ETHERNET_LINK_1000BASE_T | ETHERNET_LINK_10BASE_T;
+}
+
+static void eth_nxp_enet_qos_rx(struct k_work *work)
+{
+	struct nxp_enet_qos_rx_data *rx_data =
+		CONTAINER_OF(work, struct nxp_enet_qos_rx_data, rx_work);
+	struct nxp_enet_qos_mac_data *data =
+		CONTAINER_OF(rx_data, struct nxp_enet_qos_mac_data, rx);
+	volatile union nxp_enet_qos_rx_desc *desc_arr = rx_data->descriptors;
+	volatile union nxp_enet_qos_rx_desc *desc;
+	struct net_pkt *pkt;
+	struct net_buf *new_buf;
+	struct net_buf *buf;
+	size_t pkt_len;
+
+	/* We are going to find all of the descriptors we own and update them */
+	for (int i = 0; i < NUM_RX_BUFDESC; i++) {
+		desc = &desc_arr[i];
+
+		if (desc->write.control3 & OWN_FLAG) {
+			/* The DMA owns the descriptor, we cannot touch it */
+			continue;
+		}
+
+		/* Otherwise, we found a packet that we need to process */
+		pkt = net_pkt_rx_alloc(K_NO_WAIT);
+
+		if (!pkt) {
+			LOG_ERR("Could not alloc RX pkt");
+			goto error;
+		}
+
+		LOG_DBG("Created RX pkt %p", pkt);
+
+		/* We need to know if we can replace the reserved fragment in advance.
+		 * At no point can we allow the driver to have less the amount of reserved
+		 * buffers it needs to function, so we will not give up our previous buffer
+		 * unless we know we can get a new one.
+		 */
+		new_buf = net_pkt_get_frag(pkt, CONFIG_NET_BUF_DATA_SIZE, K_NO_WAIT);
```

Comment:
FYI referencing `CONFIG_NET_BUF_DATA_SIZE` prevents enabling `CONFIG_NET_BUF_VARIABLE_DATA_SIZE=y`- the fixed allocator can be less memory efficient when there's a mix of tiny (DHCP) and large (TLS) packets.

Reply:
okay, I will add a kconfig dependency on CONFIG_NET_BUF_FIXED_DATA_SIZE until this optimization is able to be made